Minor cleanup, FreeBSD fixes and functionality, NetBSD functionality

battleblow opened the pull request.

battleblow committed:
Fix compilation on NetBSD

* Include <> for the definition of uvmexp_sysctl.
* On NetBSD 8.0, KERN_CPTIME is not defined in any header, instead
  use sysctlbyname and kern.cp_time similar to FreeBSD.

devnexen committed:
Hotspot/cpu context switch for OpenBSD

bsdkurt committed:
Additional BSD support in os_perf_bsd.cpp:

* Add support for OpenBSD and NetBSD in most functions
* Share code with APPLE and other BSDs where possible
* Use correct datatype for sysctl mib length and calculate value instead
  of hard code throughout

devnexen committed:
NetBSD context switch support

bsdkurt committed:
Minor corrections:
* fix compilation due to miss-matched #endif
* use u_int for miblen

battleblow committed:
Implement getting the system processes on NetBSD

* Use the appropriate sysctls for NetBSD to populate the process list.  The
  code structure is very similar to FreeBSD and should likely be refactored
  once the final shape of the OpenBSD version is clear.

Testing:

* Extracted code and ran in a small standalone program.
* Compile tested.

bsdkurt committed:
Adjustments from review:
* Adjust header includes
* OpenBSD/cpu_load_total_process(): remove CP_SPIN from used_ticks
* Remove duplicated miblen line

bsdkurt committed:
Revert initial implementations of cpu_load_total_process() and
cpu_loads_process() in os_perf_bsd.cpp after review.

bsdkurt committed:
JFR CPUPerformance BSD implementation:
* Implementation for *BSD that is consistent with Linux's implementation
  per discussion in #37
* cpu_load() for individual processors left to do for FreeBSD and NetBSD

devnexen committed:
JFR CPUPerformance, FreeBSD subpart (per cpu load)

devnexen committed:
Allocations fixes

devnexen committed:
JFR CPUPerformance : NetBSD per cpu load (#2)

* JFR CPUPerformance : NetBSD per cpu load
Author: "David CARLIER" <devnexen@gmail.com>

bsdkurt committed:
Adjust BSD CPUPerformanceInterface:
* Incorporate review feedback from #39
* Separate BSD implementations from Apple where there was no code
  sharing.

bsdkurt committed:
Adjustments from review.

battleblow committed:
Convert hardcoded values to constants

* Add constants for the number of microseconds per seconds and number
  of nanoseconds per microsecond and user these in preference to hard
  coded values.
* Make the constants long, but convert to a double where needed.
* Re-order and bracket a calculation for clarity.

battleblow committed:
Count all processors

* On FreeBSD, the number of active processors for the JVM may be set to an
  arbitrary set of CPUs using cpuset(1).  The current implementation of
  determining the CPU load does not take this into account and may be
  incorrect.  Also, since this can be done dynamically, the set of CPUs
  may change while the JVM is running, which would introduce other
  inconsistencies.  For now, don't attempt to deal with this but instead
  just count the load for all processors.  This will at least yield
  consistent cpu loads in this situation.

  This should have no effect on OpenBSD or NetBSD since they always have
  all processors counted as active.

battleblow committed:
Only add actual processes to SystemProcess on FreeBSD

* Switch from KERN_PROC_ALL to KERN_PROC_PROC on FreeBSD.  The latter
  only lists processes, while the former enumerates each thread, which
  means there will be duplicate entries for multi-threaded processes.

Testing:

* Extracted code and ran in a small standalone programme.
* Ran flight recorder and examined raw recording.

battleblow committed:
Add support for system process command lines on FreeBSD

* Use KERN_PROC_ARGS to extract the command line for the process using
  sysctl(3).
* Small refactoring to put some checks on the pathname earlier to avoid
  looking up the arguments if the process is going to be skipped anyway.
* Made the indenting consistent.

Testing:

* Extracted code and ran in a small standalone programme.
* Ran flight recorder and examined raw recording.

battleblow committed:
Merge branch 'bsd-port' into bsd-port-os-perf-bsd

battleblow commented:
Github seems confused about what commits are in this pull request.  I think perhaps due to the squashing of the commits going into bsd-port.

The only changes that are under review here are the commits I did today.  The earlier changes have already been merged back.  I think the right thing to do after review is for me to manually merge rather than use the UI to merge as a result.

battleblow commented:
@devnexen the UI isn't allowing me to add you as a reviewer for some reason, but I'd appreciate your review too.

devnexen commented on src/hotspot/os/bsd/os_perf_bsd.cpp:
```diff
@@ -690,26 +690,47 @@ int SystemProcessInterface::SystemProcesses::system_processes(SystemProcess** sy
+      continue;
+    }
+    char *path = NEW_C_HEAP_ARRAY(char, plen + 1, mtInternal);
+    strncpy(path, pbuf, plen);
```
does not have to change but since we re under BSD `strlcpy` might do ?

battleblow replied:
> does not have to change but since we re under BSD `strlcpy` might do ?

Good call.  Coming in a future pull request :)

battleblow commented:
I just pushed in the NetBSD implementation for system processes too.

bsdkurt commented on src/hotspot/os/bsd/os_perf_bsd.cpp:
```diff
@@ -50,7 +50,9 @@
-static const double NANOS_PER_SEC = 1000000000.0;
+static const long NANOS_PER_SEC = 1000000000L;
+static const long MICROS_PER_SEC = 1000000L;
+static const long NANOS_PER_MICROSEC = 1000L;
```
While reviewing the datatypes here I noted a few items that can be resolved in different ways. On OpenBSD/i386 long's are 4 bytes (I think also OS X/32bit). Which means `now_in_nanos()` has the 2038 problem on OpenBSD/i386. I don't think this diff needs to address that problem.

I'm not sure why the original author used a static value for NANOS_PER_SEC. I think these values would be better off as `#define` with `LL` instead of static const values. Alternatively, `static const time_t` with `LL` would also address the sizeof(long) issue here I believe.

I also noticed something else tangential to your diff, `now_in_nanos()` uses `gettimeofday(2)` which is susceptible to time moving backwards due to NTP or a sys-admin changing the time. I think we should be using `clock_gettime(CLOCK_MONOTONIC, ...)` but this can be done in a separate step if you want to get this diff in without addressing that.

devnexen replied:
Agreed on the last point, that is a better option.

bsdkurt reviewed (changes requested):
I have a minor datatype change request to the constants and noted some issues tangential to your diff regarding `now_in_nanos()` that can be addressed separately. We may also want to consider moving all uses of `malloc` to `NEW_C_HEAP_ARRAY` as we're the only os_perf_xxx.cpp implementation using malloc.

battleblow committed:
Convert time constants, use correct memory management macros

* Convert constants such as NANOS_PER_SEC to be time_t.
* Use NEW_C_HEAP_ARRAY/FREE_C_HEAP_ARRAY instead of malloc/free per the
  exhortation in allocation.hpp.  While doing this add some checks for
  allocation success that were missing.

bsdkurt reviewed (approved):
LGTM. Thanks!

battleblow commented:
I've cherry picked the relevant commits and squashed them to a single commit in bsd-port.  Merging would have brought in the ones Kurt had previously squashed as well.

What do you both think of deleting this branch and recreating it so it's cleaner wrt bsd-port?

bsdkurt commented:
> What do you both think of deleting this branch and recreating it so it's cleaner wrt bsd-port?

Yes. Sounds like a good idea.
